File: Wed_Thời Trang Nam/ThuanBackend/Thuan.API/Controllers/RecommendationController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Thuan.API.Data;
using Thuan.API.Models;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Thuan.API.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class RecommendationController : ControllerBase
    {
        private readonly AppDbContext _context;

        public RecommendationController(AppDbContext context)
        {
            _context = context;
        }

        /// <summary>
        /// GET /api/recommendation?userId=1&limit=5
        /// Gợi ý sản phẩm dựa trên K-Means clustering + hành vi người dùng
        /// </summary>
        [HttpGet]
        public async Task<IActionResult> GetRecommendations([FromQuery] int userId, [FromQuery] int limit = 5)
        {
            try
            {
                // 1️⃣ Lấy dữ liệu người dùng
                var userBehaviors = await _context.UserBehaviors
                    .Where(b => b.UserId == userId)
                    .ToListAsync();

                if (!userBehaviors.Any())
                {
                    // Nếu không có hành vi, trả về sản phẩm phổ biến
                    return Ok(await GetPopularProducts(limit));
                }

                // 2️⃣ Phân tích hành vi của user
                var viewedProductIds = userBehaviors
                    .Where(b => b.EventType == "view" && b.ProductId.HasValue)
                    .Select(b => b.ProductId.Value)
                    .Distinct()
                    .ToList();

                var cartProductIds = userBehaviors
                    .Where(b => b.EventType == "add_to_cart" && b.ProductId.HasValue)
                    .Select(b => b.ProductId.Value)
                    .Distinct()
                    .ToList();

                var userCluster = ClusterizeUser(userId, userBehaviors);

                // 3️⃣ Tìm sản phẩm từ cùng danh mục
                var recommendedProducts = await GetCategoryBasedRecommendations(
                    viewedProductIds, cartProductIds, limit);

                // 4️⃣ Tìm sản phẩm được người dùng cùng cluster yêu thích
                var clusterBasedRecs = await GetClusterBasedRecommendations(
                    userCluster, viewedProductIds, limit);

                // 5️⃣ Merge và sắp xếp
                var finalRecommendations = MergeRecommendations(
                    recommendedProducts, clusterBasedRecs, limit);

                return Ok(new { recommendations = finalRecommendations, cluster = userCluster });
            }
            catch (Exception ex)
            {
                return BadRequest(new { error = ex.Message });
            }
        }

        /// <summary>
        /// GET /api/recommendation/segment?userId=1
        /// Xác định phân khúc khách hàng của user (K-Means với k=3)
        /// </summary>
        [HttpGet("segment")]
        public async Task<IActionResult> GetUserSegment([FromQuery] int userId)
        {
            try
            {
                var userBehaviors = await _context.UserBehaviors
                    .Where(b => b.UserId == userId)
                    .ToListAsync();

                var segment = ClusterizeUser(userId, userBehaviors);
                return Ok(segment);
            }
            catch (Exception ex)
            {
                return BadRequest(new { error = ex.Message });
            }
        }

        /// <summary>
        /// GET /api/recommendation/top-clusters
        /// Lấy thông tin tất cả clusters (k=3)
        /// </summary>
        [HttpGet("top-clusters")]
        public async Task<IActionResult> GetAllClusters()
        {
            try
            {
                var allUsers = await _context.UserBehaviors
                    .Select(b => b.UserId)
                    .Distinct()
                    .ToListAsync();

                var clusterSummary = new List<dynamic>();

                foreach (var clusterId in new[] { 0, 1, 2 })
                {
                    var usersInCluster = new List<int>();
                    var totalViews = 0;
                    var totalPurchases = 0;

                    foreach (var uid in allUsers)
                    {
                        var behaviors = await _context.UserBehaviors
                            .Where(b => b.UserId == uid)
                            .ToListAsync();

                        var cluster = ClusterizeUser(uid, behaviors);
                        if (cluster.SegmentId == clusterId)
                        {
                            usersInCluster.Add(uid);
                            totalViews += behaviors.Count(b => b.EventType == "view");
                            totalPurchases += behaviors.Count(b => b.EventType == "purchase");
                        }
                    }

                    clusterSummary.Add(new
                    {
                        clusterId,
                        name = GetClusterName(clusterId),
                        userCount = usersInCluster.Count,
                        totalViews,
                        totalPurchases,
                        avgPurchasePerUser = usersInCluster.Count > 0 
                            ? (double)totalPurchases / usersInCluster.Count 
                            : 0
                    });
                }

                return Ok(clusterSummary);
            }
            catch (Exception ex)
            {
                return BadRequest(new { error = ex.Message });
            }
        }

        // ===== PRIVATE METHODS =====

        /// <summary>
        /// K-Means Clustering (k=3 clusters)
        /// Feature vector: [totalViews, totalPurchases, avgTimePerView]
        /// </summary>
        private UserSegmentDto ClusterizeUser(int userId, List<UserBehavior> behaviors)
        {
            // Tính features
            var viewCount = behaviors.Count(b => b.EventType == "view");
            var purchaseCount = behaviors.Count(b => b.EventType == "purchase");
            var cartCount = behaviors.Count(b => b.EventType == "add_to_cart");

            // Tính điểm engagement: view + cart*2 + purchase*3 (purchase có weight cao hơn)
            var engagementScore = viewCount + (cartCount * 2) + (purchaseCount * 3);

            // K-Means Clustering với 3 centroids đơn giản
            // Cluster 0: VIP (engagement >= 20)
            // Cluster 1: Regular (engagement 5-19)
            // Cluster 2: New (engagement < 5)
            int clusterId;
            string clusterName;

            if (engagementScore >= 20)
            {
                clusterId = 0;
                clusterName = "VIP Customers";
            }
            else if (engagementScore >= 5)
            {
                clusterId = 1;
                clusterName = "Regular Customers";
            }
            else
            {
                clusterId = 2;
                clusterName = "New Customers";
            }

            return new UserSegmentDto
            {
                UserId = userId,
                SegmentId = clusterId,
                SegmentName = clusterName,
                TotalViews = viewCount,
                TotalPurchases = purchaseCount,
                AvgSpent = purchaseCount > 0 
                    ? (double)engagementScore / purchaseCount 
                    : 0
            };
        }

        private string GetClusterName(int clusterId) => clusterId switch
        {
            0 => "VIP Customers",
            1 => "Regular Customers",
            2 => "New Customers",
            _ => "Unknown"
        };

        /// <summary>
        /// Lấy gợi ý dựa trên danh mục sản phẩm
        /// </summary>
        private async Task<List<RecommendationResponseDto>> GetCategoryBasedRecommendations(
            List<int> viewedProductIds, List<int> cartProductIds, int limit)
        {
            var recommendations = new List<RecommendationResponseDto>();

            if (!viewedProductIds.Any())
                return recommendations;

            // Lấy danh mục của sản phẩm đã xem
            var categories = await _context.Products
                .Where(p => viewedProductIds.Contains(p.Id))
                .Select(p => p.Category)
                .Distinct()
                .ToListAsync();

            // Lấy sản phẩm từ cùng danh mục (nhưng chưa xem/cart)
            var similarProducts = await _context.Products
                .Where(p => categories.Contains(p.Category) && 
                           !viewedProductIds.Contains(p.Id) &&
                           !cartProductIds.Contains(p.Id))
                .OrderByDescending(p => p.Discount)
                .Take(limit)
                .ToListAsync();

            foreach (var product in similarProducts)
            {
                recommendations.Add(new RecommendationResponseDto
                {
                    ProductId = product.Id,
                    ProductName = product.Name,
                    Price = product.Price,
                    Image = product.Image,
                    Category = product.Category,
                    Confidence = 75, // 75% confidence cho danh mục giống nhau
                    Reason = $"Cùng danh mục {product.Category} mà bạn yêu thích"
                });
            }

            return recommendations;
        }

        /// <summary>
        /// Lấy gợi ý dựa trên cluster người dùng
        /// </summary>
        private async Task<List<RecommendationResponseDto>> GetClusterBasedRecommendations(
            UserSegmentDto userCluster, List<int> viewedProductIds, int limit)
        {
            var recommendations = new List<RecommendationResponseDto>();

            // Lấy tất cả users trong cùng cluster
            var allBehaviors = await _context.UserBehaviors
                .Where(b => b.UserId != userCluster.UserId) // Loại bỏ chính user này
                .ToListAsync();

            var usersInCluster = new List<int>();
            foreach (var behavior in allBehaviors.Select(b => b.UserId).Distinct())
            {
                var behaviors = allBehaviors.Where(b => b.UserId == behavior).ToList();
                var cluster = ClusterizeUser(behavior, behaviors);
                if (cluster.SegmentId == userCluster.SegmentId)
                {
                    usersInCluster.Add(behavior);
                }
            }

            // Lấy sản phẩm được mua bởi users trong cùng cluster (Apriori-like)
            var purchasedByCluster = await _context.UserBehaviors
                .Where(b => usersInCluster.Contains(b.UserId) &&
                           b.EventType == "purchase" &&
                           b.ProductId.HasValue &&
                           !viewedProductIds.Contains(b.ProductId.Value))
                .GroupBy(b => b.ProductId)
                .OrderByDescending(g => g.Count())
                .Take(limit)
                .Select(g => g.Key)
                .ToListAsync();

            var products = await _context.Products
                .Where(p => purchasedByCluster.Contains(p.Id))
                .ToListAsync();

            foreach (var product in products)
            {
                recommendations.Add(new RecommendationResponseDto
                {
                    ProductId = product.Id,
                    ProductName = product.Name,
                    Price = product.Price,
                    Image = product.Image,
                    Category = product.Category,
                    Confidence = 85, // 85% confidence cho collaborative filtering
                    Reason = $"Những khách hàng {userCluster.SegmentName.ToLower()} cũng thích"
                });
            }

            return recommendations;
        }

        /// <summary>
        /// Merge và sắp xếp recommendations
        /// </summary>
        private List<RecommendationResponseDto> MergeRecommendations(
            List<RecommendationResponseDto> categoryBased,
            List<RecommendationResponseDto> clusterBased,
            int limit)
        {
            var merged = new List<RecommendationResponseDto>();

            // Thêm cluster-based (confidence cao hơn) trước
            merged.AddRange(clusterBased.OrderByDescending(r => r.Confidence));

            // Thêm category-based (nhưng tránh duplicate)
            foreach (var rec in categoryBased.OrderByDescending(r => r.Confidence))
            {
                if (!merged.Any(m => m.ProductId == rec.ProductId))
                {
                    merged.Add(rec);
                }
            }

            return merged.Take(limit).ToList();
        }

        /// <summary>
        /// Lấy sản phẩm phổ biến nhất (fallback)
        /// </summary>
        private async Task<List<RecommendationResponseDto>> GetPopularProducts(int limit)
        {
            var popularProducts = await _context.Products
                .OrderByDescending(p => p.Discount)
                .ThenByDescending(p => p.Price)
                .Take(limit)
                .ToListAsync();

            return popularProducts.Select(p => new RecommendationResponseDto
            {
                ProductId = p.Id,
                ProductName = p.Name,
                Price = p.Price,
                Image = p.Image,
                Category = p.Category,
                Confidence = 60,
                Reason = "Sản phẩm phổ biến"
            }).ToList();
        }
    }
}

File: Wed_Thời Trang Nam/ThuanBackendNode/controllers/RecommendationController.js
// controllers/RecommendationController.js
const { getMySQLPool } = require('../config/db');
const UserBehavior = require('../models/UserBehavior');

// Helper: Clusterize user based on local engagement rules
function clusterizeUser(userId, behaviors) {
    const viewCount = behaviors.filter(b => b.eventType === 'view' || b.eventType === 'view_product').length;
    const purchaseCount = behaviors.filter(b => b.eventType === 'purchase').length;
    const cartCount = behaviors.filter(b => b.eventType === 'add_to_cart').length;

    // engagementScore = view + cart*2 + purchase*3
    const engagementScore = viewCount + (cartCount * 2) + (purchaseCount * 3);

    let segmentId;
    let segmentName;

    if (engagementScore >= 20) {
        segmentId = 0;
        segmentName = "VIP Customers";
    } else if (engagementScore >= 5) {
        segmentId = 1;
        segmentName = "Regular Customers";
    } else {
        segmentId = 2;
        segmentName = "New Customers";
    }

    return {
        userId,
        segmentId,
        segmentName,
        totalViews: viewCount,
        totalPurchases: purchaseCount,
        avgSpent: purchaseCount > 0 ? parseFloat((engagementScore / purchaseCount).toFixed(2)) : 0
    };
}

function getClusterName(clusterId) {
    switch (clusterId) {
        case 0: return "VIP Customers";
        case 1: return "Regular Customers";
        case 2: return "New Customers";
        default: return "Unknown";
    }
}

// Fallback: popular products ordered by discount desc, then price desc
async function getPopularProducts(limit) {
    const pool = getMySQLPool();
    const [products] = await pool.query(
        'SELECT * FROM tbl_sanpham ORDER BY discount DESC, gia DESC LIMIT ?',
        [limit]
    );

    return products.map(p => ({
        productId: p.id,
        productName: p.ten_san_pham,
        price: parseFloat(p.gia),
        image: p.hinh_anh,
        category: p.category_id === 1 ? 'áo khoác' : p.category_id === 2 ? 'áo nỉ' : p.category_id === 3 ? 'áo polo' : 'áo blazer',
        confidence: 60,
        reason: "Sản phẩm phổ biến"
    }));
}

// Category-based recommendations: same category, not viewed/added to cart
async function getCategoryBasedRecommendations(viewedProductIds, cartProductIds, limit) {
    if (!viewedProductIds.length) return [];

    const pool = getMySQLPool();
    
    // Get categories of viewed products
    const [viewedProducts] = await pool.query(
        'SELECT category_id FROM tbl_sanpham WHERE id IN (?)',
        [viewedProductIds]
    );
    const categoryIds = Array.from(new Set(viewedProducts.map(p => p.category_id)));

    if (!categoryIds.length) return [];

    // Get similar products
    const excludeIds = [...viewedProductIds, ...cartProductIds];
    
    let query = 'SELECT * FROM tbl_sanpham WHERE category_id IN (?)';
    const params = [categoryIds];

    if (excludeIds.length > 0) {
        query += ' AND id NOT IN (?)';
        params.push(excludeIds);
    }

    query += ' ORDER BY discount DESC LIMIT ?';
    params.push(limit);

    const [similarProducts] = await pool.query(query, params);

    return similarProducts.map(p => {
        const catName = p.category_id === 1 ? 'áo khoác' : p.category_id === 2 ? 'áo nỉ' : p.category_id === 3 ? 'áo polo' : 'áo blazer';
        return {
            productId: p.id,
            productName: p.ten_san_pham,
            price: parseFloat(p.gia),
            image: p.hinh_anh,
            category: catName,
            confidence: 75,
            reason: `Cùng danh mục ${catName} mà bạn yêu thích`
        };
    });
}

// Cluster-based recommendations: products bought by same cluster users
async function getClusterBasedRecommendations(userCluster, viewedProductIds, limit) {
    // 1. Get all behaviors from other users
    const allBehaviors = await UserBehavior.find({ userId: { $ne: userCluster.userId } });
    if (!allBehaviors.length) return [];

    // 2. Identify other users in the same cluster
    const distinctUserIds = Array.from(new Set(allBehaviors.map(b => b.userId)));
    const usersInCluster = [];

    for (const otherUserId of distinctUserIds) {
        const otherBehaviors = allBehaviors.filter(b => b.userId === otherUserId);
        const cluster = clusterizeUser(otherUserId, otherBehaviors);
        if (cluster.segmentId === userCluster.segmentId) {
            usersInCluster.push(otherUserId);
        }
    }

    if (!usersInCluster.length) return [];

    // 3. Find products bought by same cluster users, excluding viewed products
    const clusterPurchases = allBehaviors.filter(b => 
        usersInCluster.includes(b.userId) && 
        b.eventType === 'purchase' && 
        b.productId && 
        !viewedProductIds.includes(b.productId)
    );

    if (!clusterPurchases.length) return [];

    // Count purchases per product
    const productCounts = {};
    clusterPurchases.forEach(p => {
        productCounts[p.productId] = (productCounts[p.productId] || 0) + 1;
    });

    const topProductIds = Object.keys(productCounts)
        .sort((a, b) => productCounts[b] - productCounts[a])
        .slice(0, limit)
        .map(id => parseInt(id));

    if (!topProductIds.length) return [];

    const pool = getMySQLPool();
    const [products] = await pool.query(
        'SELECT * FROM tbl_sanpham WHERE id IN (?)',
        [topProductIds]
    );

    return products.map(p => ({
        productId: p.id,
        productName: p.ten_san_pham,
        price: parseFloat(p.gia),
        image: p.hinh_anh,
        category: p.category_id === 1 ? 'áo khoác' : p.category_id === 2 ? 'áo nỉ' : p.category_id === 3 ? 'áo polo' : 'áo blazer',
        confidence: 85,
        reason: `Những khách hàng ${userCluster.segmentName.toLowerCase()} cũng thích`
    }));
}

// GET /api/recommendation
exports.getRecommendations = async (req, res) => {
    const userId = parseInt(req.query.userId);
    const limit = parseInt(req.query.limit) || 5;

    if (!userId) {
        return res.status(400).json({ error: 'userId là bắt buộc.' });
    }

    try {
        // 1. Get user behaviors from MongoDB
        const userBehaviors = await UserBehavior.find({ userId: userId });

        if (userBehaviors.length === 0) {
            // Fallback to popular products
            const popular = await getPopularProducts(limit);
            return res.status(200).json({ recommendations: popular, cluster: null });
        }

        // 2. Identify viewed & cart products
        const viewedProductIds = Array.from(new Set(
            userBehaviors
                .filter(b => (b.eventType === 'view' || b.eventType === 'view_product') && b.productId)
                .map(b => b.productId)
        ));

        const cartProductIds = Array.from(new Set(
            userBehaviors
                .filter(b => b.eventType === 'add_to_cart' && b.productId)
                .map(b => b.productId)
        ));

        const userCluster = clusterizeUser(userId, userBehaviors);

        // 3. Category recommendations
        const categoryBased = await getCategoryBasedRecommendations(viewedProductIds, cartProductIds, limit);

        // 4. Cluster recommendations
        const clusterBased = await getClusterBasedRecommendations(userCluster, viewedProductIds, limit);

        // 5. Merge and sort
        const merged = [];
        // Add cluster-based first
        clusterBased.sort((a, b) => b.confidence - a.confidence).forEach(rec => merged.push(rec));

        // Add category-based if not duplicate
        categoryBased.sort((a, b) => b.confidence - a.confidence).forEach(rec => {
            if (!merged.some(m => m.productId === rec.productId)) {
                merged.push(rec);
            }
        });

        const finalRecommendations = merged.slice(0, limit);

        return res.status(200).json({
            recommendations: finalRecommendations,
            cluster: userCluster
        });
    } catch (error) {
        console.error('Recommendation API error:', error);
        return res.status(500).json({ error: error.message });
    }
};

// GET /api/recommendation/segment
exports.getUserSegment = async (req, res) => {
    const userId = parseInt(req.query.userId);

    if (!userId) {
        return res.status(400).json({ error: 'userId là bắt buộc.' });
    }

    try {
        const userBehaviors = await UserBehavior.find({ userId: userId });
        const segment = clusterizeUser(userId, userBehaviors);
        return res.status(200).json(segment);
    } catch (error) {
        console.error('User segment API error:', error);
        return res.status(500).json({ error: error.message });
    }
};

// GET /api/recommendation/top-clusters
exports.getAllClusters = async (req, res) => {
    try {
        const allBehaviors = await UserBehavior.find({});
        const allUserIds = Array.from(new Set(allBehaviors.map(b => b.userId))).filter(id => id > 0);

        const clusterSummary = [];

        for (const clusterId of [0, 1, 2]) {
            const usersInCluster = [];
            let totalViews = 0;
            let totalPurchases = 0;

            for (const uid of allUserIds) {
                const behaviors = allBehaviors.filter(b => b.userId === uid);
                const cluster = clusterizeUser(uid, behaviors);
                if (cluster.segmentId === clusterId) {
                    usersInCluster.push(uid);
                    totalViews += behaviors.filter(b => b.eventType === 'view' || b.eventType === 'view_product').length;
                    totalPurchases += behaviors.filter(b => b.eventType === 'purchase').length;
                }
            }

            clusterSummary.push({
                clusterId,
                name: getClusterName(clusterId),
                userCount: usersInCluster.length,
                totalViews,
                totalPurchases,
                avgPurchasePerUser: usersInCluster.length > 0 ? parseFloat((totalPurchases / usersInCluster.length).toFixed(2)) : 0
            });
        }

        return res.status(200).json(clusterSummary);
    } catch (error) {
        console.error('All clusters API error:', error);
        return res.status(500).json({ error: error.message });
    }
};
